plugins/inputs/libvirtd/libvirtd.go

```diff
@@ -0,0 +1,194 @@
+
+	var diskDevices []string
+
+	rXML, err := l.client.DomainGetXMLDesc(dom, 0)
+
+	if err != nil {
+		acc.AddError(fmt.Errorf("error getting domain xml dump: %+v", err))
+		return
+	}
+
+	doc := etree.NewDocument()
+	err = doc.ReadFromString(rXML)
+
+	if err != nil {
+		acc.AddError(fmt.Errorf("error reading dumpxml data: %+v", err))
+		return
+	}
+
+	domainInfo := doc.SelectElement("domain")
+
+	if domainInfo == nil {
+		acc.AddError(fmt.Errorf("domain info nil for %+v", domainInfo.SelectAttr("name").Value))
+		return
+	}
+
+	deviceElements := domainInfo.SelectElements("devices")
+
+	for _, device := range deviceElements {
+		if device.Tag == "devices" {
+			for _, disk := range device.SelectElements("disk") {
+				if disk.SelectElement("source") != nil {
+					diskDevices = append(diskDevices, disk.SelectElement("target").SelectAttr("dev").Value)
+				}
+			}
+		}
+	}
+
+	l.DiskDevices = diskDevices
+
+}
+
+func (l *LibVirtD) gatherVMDiskData(dom libvirt.Domain, acc telegraf.Accumulator) {
+
+	l.getDiskDevices(dom, acc)
+
+	for _, d := range l.DiskDevices {
+		tags := make(map[string]string)
+		fields := make(map[string]interface{})
+
+		tags["device"] = d
+		if blkParams, _, err := l.client.DomainBlockStatsFlags(dom, d, 8, 0); err == nil {
+			for _, blk := range blkParams {
+				fields[blk.Field] = blk.Value.Get()
+			}
+		} else {
+			acc.AddError(fmt.Errorf("unable to get block device data for: %v: %v", d, err))
+		}
+
+		rAlloc, rCapacity, rPhysical, err := l.client.DomainGetBlockInfo(dom, d, 0)
+		if err != nil {
+			acc.AddError(fmt.Errorf("failed to get block info for %s on domain %s it may not be available. Error: %v", d, dom.Name, err))
+			return
+		}
+
+		fields["allocation"] = rAlloc
+		fields["capacity"] = rCapacity
+		fields["physical"] = rPhysical
+		tags["name"] = dom.Name
+
+		acc.AddFields("livbirtd_disk", fields, tags)
+
+	}
+}
+
+func (l *LibVirtD) gatherVMData(dom libvirt.Domain, acc telegraf.Accumulator) {
+
+	fields := make(map[string]interface{})
+	tags := make(map[string]string)
+
+	perfEvents, err := l.client.DomainGetPerfEvents(dom, 0)
+
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+
+	for _, p := range perfEvents {
+		fields[p.Field] = p.Value.Get()
+	}
+
+	state, maxMem, currentMemory, cpuTotal, cpuTime, err := l.client.DomainGetInfo(dom)
+
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+
+	cpuParams, _, err := l.client.DomainGetCPUStats(dom, 2, 0, uint32(cpuTotal), 0)
+
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+
+	for _, c := range cpuParams {
+		fields[c.Field] = c.Value.Get()
+	}
+
+	usedMem := maxMem / currentMemory
+
+	fields["mem_used_percent"] = usedMem
+	fields["state"] = state
+	fields["max_memory"] = maxMem
+	fields["total_memory"] = maxMem
+	fields["cpu_count"] = cpuTotal
+	fields["cpu_time"] = cpuTime
+
+	tags["name"] = dom.Name
+	tags["id"] = string(dom.ID)
+
+	acc.AddFields("libvirtd_vmdata", fields, tags)
```

Comment:
I would call this `libvirt_domain`, we want to stick to the upstream terminology if possible.